src/replication/stream_senders_manager.cpp

```diff
@@ -0,0 +1,244 @@
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * stream_senders_manager.cpp -
+ */
+
+#include "stream_senders_manager.hpp"
+#include "replication_common.hpp"
+
+#include <utility>
+#include "thread_manager.hpp"
+#include "thread_daemon.hpp"
+
+/* TODO : I am not sure we will not need some other functionality specific to replication
+ * keep this in replication namespace and folder until it is more clear */
+namespace cubreplication
+{
+
+  const unsigned int stream_senders_manager::SUPERVISOR_DAEMON_DELAY_MS = 10;
+  const unsigned int stream_senders_manager::SUPERVISOR_DAEMON_CHECK_CONN_MS = 5000;
+
+  stream_senders_manager::stream_senders_manager (cubstream::stream &supervised_stream):
+    m_stream (supervised_stream)
+  {
+    init ();
+  }
+
+  stream_senders_manager::~stream_senders_manager ()
+  {
+    finalize ();
+  }
+
+  void stream_senders_manager::init ()
+  {
+#if defined (SERVER_MODE)
+    int error_code = NO_ERROR;
+
+    stream_senders.clear ();
+
+    std::string daemon_name = "senders_supervisor_daemon_" + m_stream.name ();
+    auto exec_f = std::bind (&stream_senders_manager::execute, this, std::placeholders::_1);
+    cubthread::entry_callable_task *task = new cubthread::entry_callable_task (exec_f);
+    supervisor_daemon = cubthread::get_manager ()->create_daemon (
+				cubthread::looper (std::chrono::milliseconds (SUPERVISOR_DAEMON_DELAY_MS)), task, daemon_name.c_str ());
+
+    error_code = rwlock_initialize (&senders_lock, "MASTER_SENDERS_LOCK");
+    assert (error_code == NO_ERROR);
+#endif
+  }
+
+  void stream_senders_manager::add_stream_sender (cubstream::transfer_sender *sender)
+  {
+    rwlock_write_lock (&senders_lock);
+    stream_senders.push_back (sender);
+    rwlock_write_unlock (&senders_lock);
+  }
+
+  void stream_senders_manager::stop_stream_sender (cubstream::transfer_sender *sender)
+  {
+    rwlock_write_lock (&senders_lock);
+    for (cubstream::transfer_sender *s : stream_senders)
+      {
+	if (sender == s)
+	  {
+	    er_log_debug_replication (ARG_FILE_LINE, "stream_senders_manager::stop_stream_sender for channel:%s",
+				      s->get_channel ().get_channel_id ().c_str ());
+	    s->enter_termination_phase ();
+	    break;
+	  }
+      }
+    rwlock_write_unlock (&senders_lock);
+  }
+
+
+  bool stream_senders_manager::find_stream_sender (const cubstream::transfer_sender *sender)
```

Comment:
unused find_stream_sender

Reply:
Will be used in future PR; I preferred to include it in this PR.